proxy/src/main/java/com/velocitypowered/proxy/command/SuggestionsProvider.java

```diff
@@ -0,0 +1,367 @@
+   *
+   * @param node the node to parse
+   * @param originalReader the input reader
+   * @param contextSoFar the context, containing the alias node of the command
+   * @return the parse results containing the parsed hint nodes
+   * @see VelocityCommandMeta#copyHints(CommandMeta) for the conditions under which the returned
+   *      hints can be suggested to a {@link CommandSource}.
+   */
+  private ParseResults<S> parseHints(final CommandNode<S> node, final StringReader originalReader,
+                                     final CommandContextBuilder<S> contextSoFar) {
+    // This is a stripped-down version of CommandDispatcher#parseNodes that doesn't
+    // check the requirements are satisfied and ignores redirects, neither of which
+    // are used by hint nodes. Parsing errors are ignored.
+    List<ParseResults<S>> potentials = null;
+    for (final CommandNode<S> child : node.getRelevantNodes(originalReader)) {
+      if (VelocityCommands.isArgumentsNode(child)) {
+        continue;
+      }
+      final CommandContextBuilder<S> context = contextSoFar.copy();
+      final StringReader reader = new StringReader(originalReader);
+      try {
+        // We intentionally don't catch all unchecked exceptions
+        child.parse(reader, context);
+        if (reader.canRead() && reader.peek() != CommandDispatcher.ARGUMENT_SEPARATOR_CHAR) {
+          continue;
+        }
+      } catch (final CommandSyntaxException e) {
+        continue;
+      }
+      if (reader.canRead(2)) { // separator + string
+        reader.skip(); // separator
+        final ParseResults<S> parse = this.parseHints(child, reader, context);
+        if (potentials == null) {
+          potentials = new ArrayList<>(1);
+        }
+        potentials.add(parse);
+      }
+    }
+    if (potentials != null) {
+      if (potentials.size() > 1) {
+        potentials.sort((a, b) -> {
+          if (!a.getReader().canRead() && b.getReader().canRead()) {
+            return -1;
+          }
+          if (a.getReader().canRead() && !b.getReader().canRead()) {
+            return 1;
+          }
+          return 0;
+        });
+      }
+      return potentials.get(0);
+    }
+    return new ParseResults<>(contextSoFar, originalReader, Collections.emptyMap());
+  }
+
+  /**
+   * Returns a future that is completed with the result of merging the {@link Suggestions}
+   * the given futures complete with. The results of the futures that complete exceptionally
+   * are ignored.
+   *
+   * @param fullInput the command input
+   * @param futures the futures that complete with the suggestions
+   * @return the future that completes with the merged suggestions
+   */
+  @SafeVarargs
+  private CompletableFuture<Suggestions> merge(
+          final String fullInput, final CompletableFuture<Suggestions>... futures) {
+    // https://github.com/Mojang/brigadier/pull/81
+    return CompletableFuture.allOf(futures).handle((unused, throwable) -> {
```

Comment:
Consider using [Spotify `completable-futures`](https://github.com/spotify/completable-futures) which Velocity already includes. in particular, `CompletableFutures.combineFutures()` would probably lead to a cleaner implementation of this.

Reply:
Unfortunately `completable-futures` only provides overloads of [up to 6 `CompletableFuture`s](https://javadoc.io/doc/com.spotify/completable-futures/latest/index.html) with no varargs variant. However, e.g. `provideAliasSuggestions` needs to merge an arbitrary number of futures. In addition, the function uses [`thenCompose`](https://github.com/spotify/completable-futures/blob/bfd73772ebbf7ee60f742ebb4f176830ab1b9345/src/main/java/com/spotify/futures/CompletableFutures.java#L574), which is only called if all the futures complete normally. However, we also want to merge the suggestions (skipping the ones that completed exceptionally) on exceptional completion of the `allOf` future.